File: scripts/scrape-product-content.ts
import { config } from 'dotenv';
import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';
import * as path from 'path';

config({ path: '.env.local' });

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseServiceKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !supabaseServiceKey) {
  console.error('Missing Supabase credentials');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseServiceKey, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
});

interface ScrapedData {
  images: string[];
  audioFiles: string[];
  videoLinks: string[];
  description: string;
  shortDescription: string;
  features: string[];
  galleryImages: string[];
}

async function scrapeProductPage(slug: string): Promise<ScrapedData | null> {
  const url = `https://nnaud.io/product/${slug}/`;
  
  console.log(`\n${'='.repeat(60)}`);
  console.log(`Scraping: ${slug}`);
  console.log(`URL: ${url}`);
  console.log(`${'='.repeat(60)}`);
  
  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
      }
    });
    
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }
    
    const html = await response.text();
    
    const data: ScrapedData = {
      images: [],
      audioFiles: [],
      videoLinks: [],
      description: '',
      shortDescription: '',
      features: [],
      galleryImages: []
    };
    
    // Extract images using regex - more comprehensive
    const imgRegex = /<img[^>]+(?:src|data-src|data-lazy-src|data-large_image|data-original)=["']([^"']+)["'][^>]*>/gi;
    const imgMatches = [...html.matchAll(imgRegex)];
    
    const allImageUrls = new Set<string>();
    
    for (const match of imgMatches) {
      let imgSrc = match[1];
      if (imgSrc) {
        // Clean up the URL
        imgSrc = imgSrc.replace(/^\/\//, 'https://');
        if (!imgSrc.startsWith('http')) {
          imgSrc = new URL(imgSrc, url).href;
        }
        // Normalize URL
        try {
          const urlObj = new URL(imgSrc);
          imgSrc = urlObj.href;
          if (imgSrc && !allImageUrls.has(imgSrc)) {
            allImageUrls.add(imgSrc);
            // Filter out common non-product images
            if (!imgSrc.includes('logo') && 
                !imgSrc.includes('icon') && 
                !imgSrc.includes('avatar') &&
                !imgSrc.includes('wp-content/themes') &&
                !imgSrc.includes('placeholder') &&
                !imgSrc.includes('spinner') &&
                !imgSrc.includes('gravatar') &&
                (imgSrc.includes('uploads') || imgSrc.includes('product'))) {
              data.images.push(imgSrc);
              console.log(`  ✓ Image: ${imgSrc.substring(0, 80)}...`);
            }
          }
        } catch (e) {
          // Invalid URL, skip
        }
      }
    }
    
    // Extract audio files from audio players (Sonaar, etc.)
    // Look for data-src or src attributes in audio elements
    const audioRegex = /<audio[^>]+(?:src|data-src)=["']([^"']+)["'][^>]*>/gi;
    const audioMatches = [...html.matchAll(audioRegex)];
    for (const match of audioMatches) {
      let audioSrc = match[1];
      if (audioSrc && !audioSrc.startsWith('http')) {
        audioSrc = new URL(audioSrc, url).href;
      }
      if (audioSrc && !data.audioFiles.includes(audioSrc)) {
        data.audioFiles.push(audioSrc);
        console.log(`  ✓ Audio: ${audioSrc}`);
      }
    }
    
    // Extract audio from Sonaar player - look for JSON data
    const sonaarDataRegex = /data-src=["']([^"']*\.(mp3|wav|ogg|m4a))["']/gi;
    const sonaarDataMatches = [...html.matchAll(sonaarDataRegex)];
    for (const match of sonaarDataMatches) {
      let audioSrc = match[1];
      if (audioSrc && !audioSrc.startsWith('http')) {
        audioSrc = new URL(audioSrc, url).href;
      }
      if (audioSrc && !data.audioFiles.includes(audioSrc)) {
        data.audioFiles.push(audioSrc);
        console.log(`  ✓ Audio (Sonaar data): ${audioSrc}`);
      }
    }
    
    // Extract audio from Sonaar player JSON configuration - try multiple patterns
    // Pattern 1: data-playlist attribute
    let sonaarJsonRegex = /iron-audioplayer[^>]*data-playlist=["']([^"']+)["']/gi;
    let sonaarJsonMatches = [...html.matchAll(sonaarJsonRegex)];
    for (const match of sonaarJsonMatches) {
      try {
        const encoded = match[1];
        const decoded = decodeURIComponent(encoded);
        const jsonData = JSON.parse(decoded);
        if (jsonData.tracks && Array.isArray(jsonData.tracks)) {
          for (const track of jsonData.tracks) {
            if (track.src) {
              let audioSrc = track.src;
              if (!audioSrc.startsWith('http')) {
                audioSrc = new URL(audioSrc, url).href;
              }
              if (audioSrc && !data.audioFiles.includes(audioSrc)) {
                data.audioFiles.push(audioSrc);
                console.log(`  ✓ Audio (Sonaar track): ${audioSrc}`);
              }
            }
          }
        }
      } catch (e) {
        // Try unescaped
        try {
          const jsonData = JSON.parse(match[1]);
          if (jsonData.tracks && Array.isArray(jsonData.tracks)) {
            for (const track of jsonData.tracks) {
              if (track.src) {
                let audioSrc = track.src;
                if (!audioSrc.startsWith('http')) {
                  audioSrc = new URL(audioSrc, url).href;
                }
                if (audioSrc && !data.audioFiles.includes(audioSrc)) {
                  data.audioFiles.push(audioSrc);
                  console.log(`  ✓ Audio (Sonaar track): ${audioSrc}`);
                }
              }
            }
          }
        } catch (e2) {
          // Not valid JSON, skip
        }
      }
    }
    
    // Pattern 2: Look for Sonaar player script initialization
    const sonaarScriptRegex = /sonaar.*?tracks.*?\[(.*?)\]/gis;
    const sonaarScriptMatches = [...html.matchAll(sonaarScriptRegex)];
    for (const match of sonaarScriptMatches) {
      // Try to extract URLs from the tracks array
      const urlRegex = /(https?:\/\/[^\s"']+\.(mp3|wav|ogg|m4a))/gi;
      const urlMatches = [...match[0].matchAll(urlRegex)];
      for (const urlMatch of urlMatches) {
        const audioSrc = urlMatch[1];
        if (audioSrc && !data.audioFiles.includes(audioSrc)) {
          data.audioFiles.push(audioSrc);
          console.log(`  ✓ Audio (Sonaar script): ${audioSrc}`);
        }
      }
    }
    
    // Look for audio in script tags (Sonaar embeds data in scripts)
    const scriptRegex = /<script[^>]*>(.*?)<\/script>/gis;
    const scriptMatches = [...html.matchAll(scriptRegex)];
    for (const match of scriptMatches) {
      const scriptContent = match[1];
      // Look for audio URLs in scripts
      const audioUrlRegex = /(https?:\/\/[^\s"']+\.(mp3|wav|ogg|m4a))/gi;
      const audioUrlMatches = [...scriptContent.matchAll(audioUrlRegex)];
      for (const urlMatch of audioUrlMatches) {
        const audioSrc = urlMatch[1];
        if (audioSrc && !data.audioFiles.includes(audioSrc)) {
          data.audioFiles.push(audioSrc);
          console.log(`  ✓ Audio (script): ${audioSrc}`);
        }
      }
    }
    
    // Extract audio links
    const audioLinkRegex = /<a[^>]+href=["']([^"']*\.(mp3|wav|ogg|m4a))["'][^>]*>/gi;
    const audioLinkMatches = [...html.matchAll(audioLinkRegex)];
    for (const match of audioLinkMatches) {
      let audioHref = match[1];
      if (audioHref && !audioHref.startsWith('http')) {
        audioHref = new URL(audioHref, url).href;
      }
      if (audioHref && !data.audioFiles.includes(audioHref)) {
        data.audioFiles.push(audioHref);
        console.log(`  ✓ Audio link: ${audioHref}`);
      }
    }
    
    // Extract video links
    const videoRegex = /<video[^>]+src=["']([^"']+)["'][^>]*>/gi;
    const videoMatches = [...html.matchAll(videoRegex)];
    for (const match of videoMatches) {
      let videoSrc = match[1];
      if (videoSrc && !videoSrc.startsWith('http')) {
        videoSrc = new URL(videoSrc, url).href;
      }
      if (videoSrc && !data.videoLinks.includes(videoSrc)) {
        data.videoLinks.push(videoSrc);
        console.log(`  ✓ Video: ${videoSrc}`);
      }
    }
    
    // Extract iframe videos (YouTube, Vimeo)
    const iframeRegex = /<iframe[^>]+src=["']([^"']+)["'][^>]*>/gi;
    const iframeMatches = [...html.matchAll(iframeRegex)];
    for (const match of iframeMatches) {
      const iframeSrc = match[1];
      if (iframeSrc && (iframeSrc.includes('youtube') || iframeSrc.includes('vimeo') || iframeSrc.includes('video'))) {
        if (!data.videoLinks.includes(iframeSrc)) {
          data.videoLinks.push(iframeSrc);
          console.log(`  ✓ Video iframe: ${iframeSrc.substring(0, 80)}...`);
        }
      }
    }
    
    // Extract description - try multiple patterns
    // Try WooCommerce short description first
    let shortDescMatch = html.match(/<div[^>]*class="[^"]*woocommerce-product-details__short-description[^"]*"[^>]*>(.*?)<\/div>/is);
    if (!shortDescMatch) {
      shortDescMatch = html.match(/<div[^>]*class="[^"]*product-short-description[^"]*"[^>]*>(.*?)<\/div>/is);
    }
    if (!shortDescMatch) {
      shortDescMatch = html.match(/<div[^>]*class="[^"]*entry-summary[^"]*"[^>]*>(.*?)<\/div>/is);
    }
    
    if (shortDescMatch) {
      data.shortDescription = shortDescMatch[1]
        .replace(/<[^>]+>/g, ' ')
        .replace(/\s+/g, ' ')
        .trim()
        .substring(0, 500);
      if (data.shortDescription) {
        console.log(`  ✓ Short description: ${data.shortDescription.length} chars`);
      }
    }
    
    // Extract full description - try multiple patterns
    // Try tab description first (most common on WooCommerce) - use non-greedy with larger capture
    let descMatch = html.match(/<div[^>]*class="[^"]*woocommerce-Tabs-panel[^"]*description[^"]*"[^>]*>(.*?)(?:<\/div>\s*<\/div>|<\/div>\s*<div[^>]*class="[^"]*woocommerce-Tabs-panel)/is);
    if (!descMatch) {
      descMatch = html.match(/<div[^>]*class="[^"]*woocommerce-product-details__description[^"]*"[^>]*>(.*?)(?:<\/div>\s*<\/div>|<\/div>\s*<div[^>]*class="[^"]*woocommerce)/is);
    }
    if (!descMatch) {
      descMatch = html.match(/<div[^>]*class="[^"]*product-description[^"]*"[^>]*>(.*?)(?:<\/div>\s*<\/div>|<\/div>\s*<div[^>]*class="[^"]*product)/is);
    }
    if (!descMatch) {
      descMatch = html.match(/<div[^>]*class="[^"]*entry-content[^"]*"[^>]*>(.*?)(?:<\/div>\s*<\/div>|<\/div>\s*<div[^>]*class="[^"]*entry)/is);
    }
    if (!descMatch) {
      descMatch = html.match(/<div[^>]*id="[^"]*product-description[^"]*"[^>]*>(.*?)(?:<\/div>\s*<\/div>|<\/div>\s*<div[^>]*id="[^"]*tab)/is);
    }
    // Also try to get content between description tab and next tab
    if (!descMatch || descMatch[1].length < 100) {
      const tabMatch = html.match(/id="tab-description"[^>]*>.*?<\/h2>(.*?)(?:<div[^>]*id="tab-|<\/div>\s*<\/div>\s*<\/div>\s*<\/div>\s*<\/div>)/is);
      if (tabMatch && tabMatch[1].length > (descMatch?.[1].length || 0)) {
        descMatch = tabMatch;
      }
    }
    
    // Extract initial description from descMatch
    let initialDescription = '';
    if (descMatch) {
      const descHtml = descMatch[1];
      // Extract paragraphs
      const pMatches = descHtml.match(/<p[^>]*>(.*?)<\/p>/gis);
      if (pMatches && pMatches.length > 0) {
        initialDescription = pMatches
          .map(m => m.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim())
          .filter(p => p.length > 10)
          .join('\n\n');
      } else {
        // Try to extract text content
        initialDescription = descHtml
          .replace(/<script[^>]*>.*?<\/script>/gis, '')
          .replace(/<style[^>]*>.*?<\/style>/gis, '')
          .replace(/<[^>]+>/g, ' ')
          .replace(/\s+/g, ' ')
          .trim();
      }
      if (initialDescription) {
        console.log(`  ✓ Description: ${initialDescription.length} chars`);
      }
    }
    
    // Start with initial description and try to find better ones
    let bestDescription = initialDescription;
    
    // Try to extract from the main content area if description is still short
    if (!bestDescription || bestDescription.length < 100) {
      const mainContentMatch = html.match(/<main[^>]*>(.*?)<\/main>/is);
      if (mainContentMatch) {
        const mainContent = mainContentMatch[1];
        // Look for paragraphs in main content
        const mainPMatches = mainContent.match(/<p[^>]*>(.*?)<\/p>/gis);
        if (mainPMatches && mainPMatches.length > 0) {
          const mainDesc = mainPMatches
            .map(m => m.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim())
            .filter(p => p.length > 20)
            .join('\n\n');
          if (mainDesc.length > bestDescription.length) {
            bestDescription = mainDesc;
            console.log(`  ✓ Description (main content): ${bestDescription.length} chars`);
          }
        }
      }
    }
    
    // Try to get any text content from the page if we still don't have a good description
    if (!bestDescription || bestDescription.length < 100) {
      // Look for any large text blocks that might be descriptions
      const textBlocks = html.match(/<p[^>]*>(.{50,})<\/p>/gis);
      if (textBlocks && textBlocks.length > 0) {
        const combinedText = textBlocks
          .map(t => t.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim())
          .filter(t => t.length > 50 && !t.includes('cookie') && !t.includes('privacy'))
          .slice(0, 5)
          .join('\n\n');
        if (combinedText.length > bestDescription.length) {
          bestDescription = combinedText;
          console.log(`  ✓ Description (text blocks): ${bestDescription.length} chars`);
        }
      }
    }
    
    // Use the best description we found
    if (bestDescription.length > 0) {
      data.description = bestDescription;
    }
    
    // Extract features (look for lists)
    const featuresMatch = html.match(/<ul[^>]*class="[^"]*product-features[^"]*"[^>]*>(.*?)<\/ul>/is);
    if (featuresMatch) {
      const liMatches = featuresMatch[1].match(/<li[^>]*>(.*?)<\/li>/gis);
      if (liMatches) {
        for (const li of liMatches) {
          const feature = li.replace(/<[^>]+>/g, '').trim();
          if (feature && feature.length > 10 && !data.features.includes(feature)) {
            data.features.push(feature);
            console.log(`  ✓ Feature: ${feature.substring(0, 50)}...`);
          }
        }
      }
    }
    
    // Gallery images (from product gallery) - get all sizes
    const galleryMatch = html.match(/<div[^>]*class="[^"]*woocommerce-product-gallery[^"]*"[^>]*>(.*?)<\/div>/is);
    if (galleryMatch) {
      const galleryImgMatches = galleryMatch[1].matchAll(imgRegex);
      for (const match of galleryImgMatches) {
        let imgSrc = match[1];
        if (imgSrc && !imgSrc.startsWith('http')) {
          imgSrc = new URL(imgSrc, url).href;
        }
        // Get the full-size image URL (remove size suffixes)
        if (imgSrc) {
          const fullSizeUrl = imgSrc.replace(/-\d+x\d+\.(webp|jpg|jpeg|png)/i, '.$1');
          if (fullSizeUrl && !data.galleryImages.includes(fullSizeUrl) && !data.images.includes(fullSizeUrl)) {
            data.galleryImages.push(fullSizeUrl);
            console.log(`  ✓ Gallery image: ${fullSizeUrl.substring(0, 80)}...`);
          }
        }
      }
    }
    
    // Also extract data-large_image attributes (WooCommerce full-size images)
    const largeImageRegex = /data-large_image=["']([^"']+)["']/gi;
    const largeImageMatches = [...html.matchAll(largeImageRegex)];
    for (const match of largeImageMatches) {
      let imgSrc = match[1];
      if (imgSrc && !imgSrc.startsWith('http')) {
        imgSrc = new URL(imgSrc, url).href;
      }
      if (imgSrc && !data.galleryImages.includes(imgSrc) && !data.images.includes(imgSrc)) {
        data.galleryImages.push(imgSrc);
        console.log(`  ✓ Large image: ${imgSrc.substring(0, 80)}...`);
      }
    }
    
    console.log(`\n  Summary:`);
    console.log(`    Images: ${data.images.length + data.galleryImages.length}`);
    console.log(`    Audio files: ${data.audioFiles.length}`);
    console.log(`    Video links: ${data.videoLinks.length}`);
    console.log(`    Features: ${data.features.length}`);
    
    return data;
    
  } catch (error) {
    console.error(`  ✗ Error scraping ${slug}:`, error);
    return null;
  }
}

async function updateProductInDb(productId: string, data: ScrapedData) {
  try {
    const updates: any = {
      updated_at: new Date().toISOString()
    };
    
    // Always update description if we have one that's better
    if (data.description && data.description.length > 0) {
      // Get current description to compare
      const { data: currentProduct } = await supabase
        .from('products')
        .select('description')
        .eq('id', productId)
        .single();
      
      const currentDescLength = currentProduct?.description?.length || 0;
      
      // Update if new description is significantly longer (at least 50 chars more)
      // or if current description is very short (< 50 chars)
      if (currentDescLength < 50 || data.description.length > currentDescLength + 50) {
        updates.description = data.description;
        console.log(`  ✓ Saving description: ${data.description.length} chars (was ${currentDescLength} chars)`);
      } else {
        console.log(`  ⚠ Keeping existing description: ${currentDescLength} chars (new: ${data.description.length} chars)`);
      }
    }
    
    if (data.shortDescription) {
      updates.short_description = data.shortDescription;
    }
    
    if (data.galleryImages.length > 0) {
      updates.gallery_images = data.galleryImages;
    }
    
    if (data.audioFiles.length > 0) {
      updates.audio_samples = data.audioFiles.map(url => ({
        url,
        name: url.split('/').pop() || 'audio'
      }));
    }
    
    if (data.videoLinks.length > 0) {
      updates.demo_video_url = data.videoLinks[0];
    }
    
    if (data.features.length > 0) {
      updates.features = data.features;
    }
    
    const { error } = await supabase
      .from('products')
      .update(updates)
      .eq('id', productId);
    
    if (error) {
      throw error;
    }
    
    console.log(`  ✓ Updated product in database`);
    return true;
  } catch (error) {
    console.error(`  ✗ Error updating database:`, error);
    return false;
  }
}

async function main() {
  const slug = process.argv[2];
  
  if (!slug) {
    console.error('Usage: npx tsx scripts/scrape-product-content.ts <slug>');
    console.error('Example: npx tsx scripts/scrape-product-content.ts 20-for-20-midi-bundle-1');
    process.exit(1);
  }
  
  // Get product ID from database
  const { data: product, error } = await supabase
    .from('products')
    .select('id')
    .eq('slug', slug)
    .single();
  
  if (error || !product) {
    console.error(`✗ Product with slug '${slug}' not found in database`);
    process.exit(1);
  }
  
  // Scrape product page
  const scrapedData = await scrapeProductPage(slug);
  
  if (!scrapedData) {
    console.error('✗ Failed to scrape product page');
    process.exit(1);
  }
  
  // Update database
  await updateProductInDb(product.id, scrapedData);
  
  console.log(`\n✅ Complete!`);
}

main().catch(console.error);

File: scripts/scrape-product-content.py
#!/usr/bin/env python3
"""
Scrape product content from nnaud.io and update database
Extracts: images, audio files, video links, descriptions, features, etc.
"""
import urllib.request
import urllib.parse
import json
import re
import sys
from html.parser import HTMLParser
from bs4 import BeautifulSoup
import os
from dotenv import load_dotenv
import psycopg2
from psycopg2.extras import Json

# Load environment variables
load_dotenv('.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL', '').replace('https://', '').replace('.supabase.co', '')
DB_PASSWORD = os.getenv('SUPABASE_DB_PASSWORD', '')
DB_HOST = f"db.{SUPABASE_URL}.supabase.co"
DB_NAME = "postgres"
DB_USER = "postgres"

def get_db_connection():
    """Get database connection"""
    return psycopg2.connect(
        host=DB_HOST,
        database=DB_NAME,
        user=DB_USER,
        password=DB_PASSWORD,
        port=5432
    )

def scrape_product_page(slug):
    """Scrape all content from a product page"""
    url = f"https://nnaud.io/product/{slug}/"
    
    print(f"\n{'='*60}")
    print(f"Scraping: {slug}")
    print(f"URL: {url}")
    print(f"{'='*60}")
    
    try:
        req = urllib.request.Request(url, headers={
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
        })
        response = urllib.request.urlopen(req, timeout=30)
        html = response.read().decode('utf-8')
        soup = BeautifulSoup(html, 'html.parser')
        
        data = {
            'images': [],
            'audio_files': [],
            'video_links': [],
            'description': '',
            'short_description': '',
            'features': [],
            'specifications': {},
            'gallery_images': []
        }
        
        # Extract main product image
        main_image = soup.find('img', class_='wp-post-image') or soup.find('img', {'data-src': re.compile('product')})
        if main_image:
            img_src = main_image.get('src') or main_image.get('data-src') or main_image.get('data-lazy-src')
            if img_src:
                if not img_src.startswith('http'):
                    img_src = urllib.parse.urljoin(url, img_src)
                data['images'].append(img_src)
                print(f"  ✓ Main image: {img_src}")
        
        # Extract all images from gallery
        gallery = soup.find('div', class_='woocommerce-product-gallery') or soup.find('div', class_='product-images')
        if gallery:
            for img in gallery.find_all('img'):
                img_src = img.get('src') or img.get('data-src') or img.get('data-lazy-src') or img.get('data-large_image')
                if img_src:
                    if not img_src.startswith('http'):
                        img_src = urllib.parse.urljoin(url, img_src)
                    if img_src not in data['gallery_images']:
                        data['gallery_images'].append(img_src)
                        print(f"  ✓ Gallery image: {img_src}")
        
        # Extract all images from content
        for img in soup.find_all('img'):
            img_src = img.get('src') or img.get('data-src') or img.get('data-lazy-src')
            if img_src and 'product' in img_src.lower():
                if not img_src.startswith('http'):
                    img_src = urllib.parse.urljoin(url, img_src)
                if img_src not in data['images'] and img_src not in data['gallery_images']:
                    data['images'].append(img_src)
        
        # Extract audio files
        for audio in soup.find_all('audio'):
            src = audio.get('src')
            if src:
                if not src.startswith('http'):
                    src = urllib.parse.urljoin(url, src)
                data['audio_files'].append(src)
                print(f"  ✓ Audio: {src}")
        
        for link in soup.find_all('a', href=re.compile(r'\.(mp3|wav|ogg|m4a)', re.I)):
            href = link.get('href')
            if href and href not in data['audio_files']:
                if not href.startswith('http'):
                    href = urllib.parse.urljoin(url, href)
                data['audio_files'].append(href)
                print(f"  ✓ Audio link: {href}")
        
        # Extract video links
        for video in soup.find_all('video'):
            src = video.get('src')
            if src:
                if not src.startswith('http'):
                    src = urllib.parse.urljoin(url, src)
                data['video_links'].append(src)
                print(f"  ✓ Video: {src}")
        
        for iframe in soup.find_all('iframe'):
            src = iframe.get('src')
            if src and ('youtube' in src or 'vimeo' in src or 'video' in src.lower()):
                data['video_links'].append(src)
                print(f"  ✓ Video iframe: {src}")
        
        # Extract description
        desc_div = soup.find('div', class_='woocommerce-product-details__short-description') or \
                   soup.find('div', class_='product-short-description') or \
                   soup.find('div', class_='entry-summary')
        if desc_div:
            data['short_description'] = desc_div.get_text(strip=True)[:500]
            print(f"  ✓ Short description: {len(data['short_description'])} chars")
        
        # Extract full description
        desc_div = soup.find('div', class_='woocommerce-product-details__description') or \
                   soup.find('div', class_='product-description') or \
                   soup.find('div', class_='entry-content')
        if desc_div:
            # Get text but preserve some formatting
            paragraphs = desc_div.find_all('p')
            data['description'] = '\n\n'.join([p.get_text(strip=True) for p in paragraphs if p.get_text(strip=True)])
            print(f"  ✓ Description: {len(data['description'])} chars")
        
        # Extract features (usually in lists)
        features_list = soup.find('ul', class_='product-features') or \
                       soup.find('ul', class_='features') or \
                       soup.find('div', class_='product-features')
        if features_list:
            for li in features_list.find_all('li'):
                feature_text = li.get_text(strip=True)
                if feature_text and feature_text not in data['features']:
                    data['features'].append(feature_text)
                    print(f"  ✓ Feature: {feature_text[:50]}")
        
        # Look for features in description
        if data['description']:
            # Look for bullet points or numbered lists
            lines = data['description'].split('\n')
            for line in lines:
                if line.strip().startswith(('•', '-', '*', '✓', '→')) or re.match(r'^\d+[\.\)]', line.strip()):
                    feature = re.sub(r'^[•\-\*\✓→\d\.\)\s]+', '', line.strip())
                    if feature and len(feature) > 10 and feature not in data['features']:
                        data['features'].append(feature)
        
        print(f"\n  Summary:")
        print(f"    Images: {len(data['images']) + len(data['gallery_images'])}")
        print(f"    Audio files: {len(data['audio_files'])}")
        print(f"    Video links: {len(data['video_links'])}")
        print(f"    Features: {len(data['features'])}")
        
        return data
        
    except Exception as e:
        print(f"  ✗ Error scraping {slug}: {e}")
        return None

def update_product_in_db(product_id, data):
    """Update product in database with scraped data"""
    conn = get_db_connection()
    try:
        cur = conn.cursor()
        
        # Prepare update query
        updates = []
        params = []
        
        if data.get('description'):
            updates.append("description = %s")
            params.append(data['description'])
        
        if data.get('short_description'):
            updates.append("short_description = %s")
            params.append(data['short_description'])
        
        if data.get('gallery_images'):
            updates.append("gallery_images = %s")
            params.append(Json(data['gallery_images']))
        
        if data.get('audio_samples'):
            updates.append("audio_samples = %s")
            params.append(Json(data['audio_samples']))
        
        if data.get('demo_video_url'):
            updates.append("demo_video_url = %s")
            params.append(data['demo_video_url'])
        
        if data.get('features'):
            updates.append("features = %s")
            params.append(Json(data['features']))
        
        if updates:
            params.append(product_id)
            query = f"""
                UPDATE products 
                SET {', '.join(updates)}, updated_at = NOW()
                WHERE id = %s
            """
            cur.execute(query, params)
            conn.commit()
            print(f"  ✓ Updated product in database")
            return True
        else:
            print(f"  ⚠ No data to update")
            return False
            
    except Exception as e:
        print(f"  ✗ Error updating database: {e}")
        conn.rollback()
        return False
    finally:
        conn.close()

def main():
    if len(sys.argv) < 2:
        print("Usage: python scrape-product-content.py <slug>")
        print("Example: python scrape-product-content.py 20-for-20-midi-bundle-1")
        sys.exit(1)
    
    slug = sys.argv[1]
    
    # Get product ID from database
    conn = get_db_connection()
    try:
        cur = conn.cursor()
        cur.execute("SELECT id FROM products WHERE slug = %s", (slug,))
        result = cur.fetchone()
        if not result:
            print(f"✗ Product with slug '{slug}' not found in database")
            sys.exit(1)
        product_id = result[0]
    finally:
        conn.close()
    
    # Scrape product page
    data = scrape_product_page(slug)
    
    if not data:
        print("✗ Failed to scrape product page")
        sys.exit(1)
    
    # Process data for database
    db_data = {}
    
    if data.get('description'):
        db_data['description'] = data['description']
    
    if data.get('short_description'):
        db_data['short_description'] = data['short_description']
    
    if data.get('gallery_images'):
        db_data['gallery_images'] = data['gallery_images']
    
    if data.get('audio_files'):
        db_data['audio_samples'] = [{'url': url, 'name': url.split('/')[-1]} for url in data['audio_files']]
    
    if data.get('video_links'):
        # Use first video as demo video
        db_data['demo_video_url'] = data['video_links'][0]
    
    if data.get('features'):
        db_data['features'] = data['features']
    
    # Update database
    update_product_in_db(product_id, db_data)
    
    print(f"\n✅ Complete!")

if __name__ == '__main__':
    main()
